src/components/nodes/boundaryEvent/boundaryEvent.vue

```diff
@@ -76,32 +81,87 @@ export default {
       this.shape.position(x - (width / 2), y - (width / 2));
       this.updateCrownPosition();
     },
-    attachBoundaryEventToTask() {
-      const task = this.getTaskUnderShape();
+    attachBoundaryEventToTask(task) {
+      const currentlyAttachedTask = this.shape.getParentCell();
 
-      if (!task) {
-        return;
+      if (currentlyAttachedTask) {
+        currentlyAttachedTask.unembed(this.shape);
       }
 
       task.embed(this.shape);
       this.node.definition.set('attachedToRef', task.component.node.definition);
       this.toggleInterruptingStyle(this.node.definition.cancelActivity);
       this.updateShapePosition(task);
+    },
+    moveBoundaryEventIfOverTask() {
+      const task = this.getTaskUnderShape();
 
-      this.shape.listenTo(this.paper, 'element:pointerup', cellView => {
+      if (!task) {
+        this.shape.position(this.validPosition.x, this.validPosition.y);
+        this.updateCrownPosition();
+        return;
+      }
+
+      if (task === this.shape.getParentCell()) {
+        this.updateShapePosition(task);
+      }
+
+      this.attachBoundaryEventToTask(task);
+    },
+    listenForValidDropTargets() {
+      this.shape.listenTo(this.paper, 'element:pointerdown', cellView => {
         if (cellView.model !== this.shape) {
           return;
         }
-        this.updateShapePosition(task);
+
+        this.validPosition = this.shape.position();
+        this.shape.listenToOnce(this.paper, 'cell:pointerup blank:pointerup', this.moveBoundaryEventIfOverTask);
+      });
+    },
+    isPoolShape(model) {
+      return model.component.node.type === 'processmaker-modeler-pool';
+    },
+    listenForInvalidDropTargets() {
+      this.shape.on('change:position', () => {
+        const targetElement = this.graph
+          .findModelsUnderElement(this.shape)
+          .filter(model => model.component)[0];
+
+        const targetHasNotChanged = this.invalidTargetElement === targetElement;
+        if (targetHasNotChanged) {
+          return;
+        }
+
+        const targetIsInvalid = targetElement && !this.isValidBoundaryEventTarget(targetElement);
+        if (targetIsInvalid) {
+          targetElement.attr('body/fill', invalidNodeColor);
+        }
+
+        if (this.invalidTargetElement) {
+          this.resetShapeColor(this.invalidTargetElement);
+        }
+
+        this.invalidTargetElement = targetElement;
       });
     },
+    resetShapeColor(shape) {
+      const defaultColor = this.isPoolShape(shape) ? poolColor : defaultNodeColor;
+      shape.attr('body/fill', defaultColor);
+    },
   },
   async mounted() {
     this.shape = new EventShape();
     this.setShapeProperties();
     this.shape.addTo(this.graph);
     await this.$nextTick();
-    this.attachBoundaryEventToTask();
+    this.listenForValidDropTargets();
+    this.listenForInvalidDropTargets();
+    this.toggleInterruptingStyle();
+
+    const task = this.getTaskUnderShape();
+    if (task) {
```

Comment:
Could this null checking be done inside `attachBoundaryEventToTask`? It would keep the code that calls this method cleaner and simplify any future changes to what it means for that method to be allowed to be called. You could use early return to bail if a task is not present.